### PR TITLE
Exclude 'tempfile' from UploadedFile's hash representation.

### LINKED ISSUE #25250: Encoding error when converting UploadedFile to json with binary tempfile
Hey,

we have an issue with `ActionDispatch::Http::UploadedFile` and `to_json`. We're using the [Gelf](https://github.com/Graylog2/gelf-rb) gem for logging to a Graylog2 instance. After (successful) API calls to a file upload action, Gelf errors in `to_json`. We traced the problem down to an instance of `UploadedFile` in Gelf's parameters hash which it tries to convert to JSON to send to the server.

The UploadedFile contains a `Tempfile` in binary mode which seems to be included in the JSON representation, including its content. The content from `read` is supposed to be encoded in `ASCII-8BIT`, yet contains binary data.

Not really sure where the best place to fix this behaviour is (Ruby's `Tempfile`, Active Support's `JSON`, or here), but I thought `UploadedFile` for sure should not include the file's content in the JSON representation. Imho.

### Steps to reproduce

Anyway, here's a test case:

https://gist.github.com/maltoe/a1b9bfc83be1c0eaef208ade9cbdfdad

### Expected behavior

Not really sure, either catch these conversion errors in `JSON` (i.e. not convert the string to UTF-8, or do character replacement in there), or simply not include the tempfile's content in the json.

### Actual behavior

```
BugTest#test_to_json_with_non_ascii_data:
Encoding::UndefinedConversionError: "\x80" from ASCII-8BIT to UTF-8
```

### Possible fix (our current monkey-patch)

```
class UploadedFile
  def to_hash
    instance_values.except 'tempfile'
  end
end
```

### System configuration

**Rails version**:
We use 4.2.5.2, but I tested on master.

**Ruby version**:
2.2.4

### PR DESCRIPTION
**Pull request by maltoe**
### Summary

This fixes `to_json`ifying an `ActionDispatch::Http::UploadedFile` when the uploaded file is binary, by
excluding the uploaded file from the object's `to_hash` representation.

Fixes #25250.

### CODE REVIEW

**Comment by sgrif** [comment]
Thanks, but this isn't a use case we want to support.

**Comment by maltoe** [comment]
Ok, but this will be an issue that will keep reappearing, I would guess. But I assume you have good reasons...

**Comment by maltoe** [comment]
Sorry that sounded a bit rude.

To clarify: I can understand that `UploadedFile` should not care about JSON serialization or serialization in general, but I don't think our use case is that much deviant from the standard, that this is to be so easily dismissed. We

 1. use lograge
 2. told lograge to include the request params in its logs
 3. use graylog2, so lograge log records have to be serialized

Basically you can replace lograge+graylog2 with any logging solution out there that sends the params back to some other service (using json, but possible other serializations). And I think it's reasonable for a gem or user to assume that it's safe to serialize the `params` to JSON, in view of they just came as form data or similar.

At which level would you suggest to fix this issue?

**Comment by sgrif** [comment]
Either your application, or the logging gem.
